### DISCUSSION TITLE
Implement Encrypted Client Hello (ECH)

### DISCUSSION THREAD

**Discussion by fabriziobagala** [mullvad/mullvadvpn-app#9076] [Feature requests & Ideas]
### I have checked if others have suggested this already

- [X] I have checked this issue tracker to see if others have reported similar issues.

### Feature description

ECH is an extension of the [TLS protocol](https://en.wikipedia.org/wiki/Transport_Layer_Security). It aims to secure the TLS handshake portion of the protocol by encrypting the name of the websites you visit, which is in plain-text. For more details, click [TLS Encrypted Client Hello](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-08).

### Alternative solutions

Currently the only solution is to disable the VPN and use a provider such as Cloudflare to get this protection.

### Type of feature

- [X] Better privacy/anonymity
- [ ] Better at circumventing censorship
- [ ] Easier to use
- [ ] Other

### Operating System

- [X] Android
- [X] iOS
- [X] Windows
- [X] macOS
- [X] Linux

**Comment by chaojoy** [comment]
Developers are strongly advised to pay attention to this issue. Even TLS1.3 will expose the server website domain name to be accessed when sending the client hello handshake data packet. This is a serious privacy threat in countries with strict network censorship and can be easily blocked by deep packet inspection systems.

**Comment by hulthe** [comment]
We are keeping an eye on this, but AFAICT we're still waiting on openssl.

**Reply by VictorVow** [reply]
This is now part of [OpenSSL 4.0](https://github.com/openssl/openssl/releases/tag/openssl-4.0.0). Is this planned to be implemented?

**Reply by barbiehunter** [reply]
> We are keeping an eye on this, but AFAICT we're still waiting on openssl.

Any update about it, as It is now part in Android 17 aswell.
https://developer.android.com/about/versions/17/summary
https://developer.android.com/about/versions/17/behavior-changes-17#ech-by-default

**Reply by hulthe** [reply]
We went the route of disabling SNI instead (#9948). AFAIK it should provide the same amount of privacy protection.
